Working with rails engines (#76)
rootstrap/blog: content/Building an Engine-modularized API in Rails - Guillermo Leopold.md

# Building an Engine-modularized API in Rails

![main image](images/engines_main_image.jpeg)

Most of the time while building an API in big development teams, one of the main challenges that we face is to have the ability to work all together in different features avoiding bottle-necks or other common issues like having a lot of conflicts or avoid overlapping between features.
Because of this, when we started working on our latest project, we decided to work with a new approach; it was also supposed to help us to scale once we need it (and we will), for example by splitting the app into microservices.

> Anyway, we knew since the beginning that it won't be easy as we are not used to work this way and there aren't too many examples around with APIs built this way. It forced us to think outside the box since the start as we couldn't start the project as we usually do, by using our Rails base project with some common gems and the signup/signin already ready to use.

Reviewer: What about `it wouldn't be easy`?

# Building an Engine-modularized API in Rails

![main image](images/engines_main_image.jpeg)

When big development teams build an API, one of the main challenges that we face is being able to work together on different features while avoiding bottle-necks or other common issues such as conflicting perspectives or avoid overlapping efforts on particular features.

> Because of this, when we started working on our latest project, we decided to work with a new approach; designed to help us to scale by splitting the app into microservices.

Reviewer 2: About this, you could add a reference or recommendation of this article:
https://martinfowler.com/bliki/MonolithFirst.html
it's about this approach we took, of modularizing an app instead of starting with microservices.

Having said that, the following is a helpful step-by-step guide on what to do when adding a new engine.

The first step, of course, is to create the engine. To achieve that, rails provides a command with some flags:
```rails plugin new engines/engine_name --mountable --skip-test --dummy-path=engines/engine_name/spec/dummy```.
Let’s got a bit deeper on how this command is built:
  * The ```--mountable``` flag means that we want to create the engine inside an isolated namespace.
  * The ```--skip-test``` flag means that we don’t want the engine to include the scaffold for the testing suite as we are going to install rspec.
  * The ```--dummy-path``` is used to define the path where we want to generate the dummy app that "works as the main app" when running the test suite.

Following this, we started by cleaning our engine, removing a lot of unnecessary files, adding some common configuration and copying some files that we need across all the engines.

> After building a few engines, we decided to improve our experience and create our own script to create the engine and also complete the following forementioned steps.

Reviewer 3: Can you add more info about this script? Maybe a minor example with the more basics things. 
An example plus the summary you provide below could be a good start point for someone that want to use engines in his projects

```Ruby
Dir.glob(File.expand_path('../engines/*', __FILE__)).each do |path|
  engine = File.basename(path)
  gem engine, path: "engines/#{engine}", require: (ENV['ENGINE'].nil? || ENV['ENGINE'] == engine)
end
```

By creating the script and adding these configurations into the main application, we were able to simplify a lot of the processes to add new engines and we ensured that all the engines are going to follow the same standard.

## Developing and testing a feature: considerations and tips

Now that we have gone through all this process that may seem extra thorough at the beginning, we are ready to start.
At this point, if everything is ok, there shouldn’t be many problems or inconveniences as we are going to be doing pretty much what we already used to do while developing features in Rails.

There are still a few considerations related to how engines work that are worth mentioning:
  * An engine works like a library, so when adding dependencies in between, we must take care to avoid circular dependencies. This means that if we have the engines A, B and C; if the engine A can see B and B can see C, then A may be able to see C but C shouldn’t be able to see either B or A.

>   * Following on from this, when adding a dependency between engines, we need to add this one into the gemfile that belongs to that engine, not the gemspec. We will also need to add the engines that the other engine has as dependencies too.

Reviewer 2: Mabye you could add some info/text about what happens when a engines that cannot see another needs to interact with it. You can talk about Observer, show some example, and maybe also recommend using async communication patterns like Queue pub/sub or similar

Author: Good one

**Cons:**
  * Lot of configuration and changes are needed when starting working with engines, requires a bit of time during the first weeks.
  * Lack of information on the internet, building an API with engines doesn’t seem to be a popular choice.
  * Despite our effort to avoid repeating code, we still repeat more code than when working in a monolith api

After overcoming the initial learning curve, we feel quite comfortable working with engines and we found out the benefits of having chosen this architecture instead of a monolith one, so I think that I would use it again if it fits the project requirements. In spite of that, we still have a lot to learn and improve upon what we already did.

As a pending feature, we would also like to see how we could take advantage of this architecture and try to deploy our application into microservices; of course that implies some changes will need to be made but it would be a valuable achievement.

## Learning about engines

To learn more about engines, the [Ruby on Rails official documentation](https://guides.rubyonrails.org/engines.html) is a good follow-on reading.

> 

Reviewer 2: can you also add a reference to that medium post in which we based (I think anthony recommended it)

# Building an Engine-modularized API in Rails

![main image](images/engines_main_image.jpeg)

When big development teams build an API, one of the main challenges that we face is being able to work together on different features while avoiding bottle-necks or other common issues such as conflicting perspectives or avoid overlapping efforts on particular features.

> Because of this, when we started working on our latest project, we decided to work with a new approach; designed to help us to scale by splitting the app into microservices. I recommend taking a look into this article to learn more about the approach we took: [Monolith First](https://martinfowler.com/bliki/MonolithFirst.html)

Reviewer: what about `help us scale`? 

(you can ignore this comment)

In our case, we were working on an e-commerce website, which made us think that some of the larger modules could be the following: catalog, user, orders, payments and admin. Of course this may grow and we are not attached to only using these modules; we may need to add new engines/modules when the app starts growing but at least this will give us the big picture of the app.

Given this, a first picture of the API architecture may look like this:

![First Glance at Engines API architecture](images/engines_first_glance.png)

It may look weird that the engines in the box are the ones that have access to the user engine and not the opposite but when implementing this, we need to have it this way.
To explain this better, let’s take the orders engine as an example:
  * The **order** engine has a model called **purchase**, it represents all the purchases in the system.
  * Each **purchase** belongs to a **customer** (which belongs to the **user** engine) and a **customer** can have many **purchases** in terms of business.

> In order to be able to model this properly, we need to add a foreign key to the customer table in the purchases table and because of this, the **purchase** (and so too the **orders**  engine) will, by definition, need to have access to the **user** engine.

Reviewer: I think this should be: `**orders**  engines`

     * The other pattern that works fine is implementing an observer pattern, in our application we followed this pattern as Rails already has a good implementation of it for ActiveRecord. This pattern works by having classes called observers that are responsible of beware of any action happening in the "observed class" and trigger actions depending on what action was taken in the followed class.

Using the **User** and **Payment** engines where **Payment** can access to **User** but **User** can't access to **Payment** to avoid having a circular dependency, let's see step by step how we implement the **observer pattern** in our application:

1. Create an observer class inside our models folder in the **Payment engine** as we want to trigger an action in this engine as a consequence of an action in the **User** engine.

2. Set up the observer in the **engine.rb** file by adding the following line:
 ```Ruby
 config.active_record.observers = :'payment/user_customer_observer'
 ```

> 3. Implement your observer, we need to specify in which model we want to observe and implement an action according that behave similar to a callback in ActiveRecord. It might look like this:

Reviewer: ```suggestion
3. Implement your observer, we need to specify which model we want to observe and implement an action according that behaves similar to a callback in ActiveRecord. It might look like this:
```

```Ruby
module Payment
  class UserCustomerObserver < ActiveRecord::Observer
    observe User::Customer
    attr_reader :customer

    def after_update(customer)
      customer.update!(payments: customer.payments + 1)
    end
  end
end
```

4. Finally, if you want to disable observers while running **rspec**, you can add this line in your test suite as a before action:
```Ruby
ActiveRecord::Base.observers.disable :all
```

> On the other hand, testing inside an engine with a rspec it’s quite similar to what we are used to doing and it provides us with some guarantees and the only considerations that took us some time are:

Reviewer: ```suggestion
On the other hand, testing inside an engine with rspec is quite similar to what we are used to doing, it provides us with some guarantees and the only considerations that took us some time are:
```

The API architecture and the project structure will look something like this:

![Engines Structure + Folder Struct](images/engines_structure_folder_struct.png)

## Summing up

It’s been a while since we started working with this project structure. Right now we are at a point where we can already look back and find the pros and cons of making this particular decision:

**Pros:**
  * Helps to parallelize work when working with medium-sized development teams.
  * Makes the code more organized and easier to read or find where a feature should be implemented.
  * The possibility to run tests inside an engine scope reduces time when testing locally.
  * A modular application like this is a mid-step between a monolith API and microservices.
  * Possibility to pick a full engine and use it in another API. We did this with our catalog engine and an admin that we develop in another project.

**Cons:**
  * Lot of configuration and changes are needed when starting working with engines, requires a bit of time during the first weeks.
  * Lack of information on the internet, building an API with engines doesn’t seem to be a popular choice.

>   * Despite our effort to avoid repeating code, we still repeat more code than when working in a monolith api

Reviewer: I would change this for: `we repeat even more code than`.

Also, small thing, the two previous paragraphs end in a `.`

After overcoming the initial learning curve, we feel quite comfortable working with engines and we found out the benefits of having chosen this architecture instead of a monolith one, so I think that I would use it again if it fits the project requirements. In spite of that, we still have a lot to learn and improve upon what we already did.

> As a pending feature, we would also like to see how we could take advantage of this architecture and try to deploy our application into microservices; of course that implies some changes will need to be made but it would be a valuable achievement.

Reviewer: What about `how we can take advantage`

## Learning about engines and references

> When we started our investigation on how to build our application, we learn a few things from this medium post: [The Modular Monolith: Rails Architecture](https://medium.com/@dan_manges/the-modular-monolith-rails-architecture-fb1023826fc4)

Reviewer: Small typo here `we learned a few things`
